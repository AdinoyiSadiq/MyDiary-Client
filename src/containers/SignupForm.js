import React, {Component} from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { signup } from '../actions/authActions';
import validateAuth from '../lib/validation';
import AuthInput from '../components/shared/AuthInput';
import Button from '../components/shared/Button';

const initialState = {
  firstname: '',
  lastname: '',
  email: '',
  password: '',
  confirmPassword: '',
  touched: {
    firstname: false,
    lastname: false,
    email: false,
    password: false,
    confirmPassword: false,
  },
};

const fieldNames = ['firstname', 'lastname', 'email', 'password', 'confirmPassword'];

class SignupForm extends Component {
	state = initialState;

	handleSubmit = async (event) => {
    event.preventDefault();
		const { signup: signupUser, history } = this.props;
		const {
      firstname, lastname, email, password, confirmPassword,
		} = this.state;
		
		const fields = {
      firstname, lastname, email, password, confirmPassword,
    };
		
		const validationError = validateAuth(fields, fieldNames);
    const changedTouchState = this.changeTouchState(fieldNames, true);
    this.setState({ touched: changedTouchState });

		const body = {
      firstName: firstname.trim(),
      lastName: lastname.trim(),
      email: email.trim(),
      password: password.trim(),
		};
		
		if (!validationError.status) {
      signupUser(body, () => history.push('/main/entries'));
      this.setState(initialState);
    }
  }

	handleChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
	}
	
	handleBlur = (field) => () => {
    const { touched } = this.state;
    this.setState({
      touched: { ...touched, [field]: true },
    });
	}
	
	changeTouchState = (fields, state) => {
    const touchState = {};
    fields.forEach((field) => {
      touchState[field] = state;
    });
    return touchState;
  }

	render() {
		const {
      firstname, lastname, email, password, confirmPassword, touched, 
		} = this.state;
		const validationError = validateAuth({
      firstname, lastname, email, password, confirmPassword,
    }, fieldNames);

		return (
			<div>
				<div className="authContent signup">
					<main>
						<div className="authLogo">MyDiary</div>
						<div className="title">Create your MyDiary account</div>
						<hr/>
						<form className="signupForm" onSubmit={this.handleSubmit}>
							<AuthInput
								value={firstname}
								touched={touched}
								error={validationError}
								handleChange={this.handleChange}
								handleBlur={this.handleBlur('firstname')} 
								name="firstname" 
								placeholder="First name" 
							/>
							<AuthInput
								value={lastname}
								touched={touched}
								error={validationError}
								handleChange={this.handleChange} 
								handleBlur={this.handleBlur('lastname')} 
								name="lastname" 
								placeholder="Last name" 
							/>
							<AuthInput
								value={email}
								touched={touched}
								error={validationError}
								handleChange={this.handleChange} 
								handleBlur={this.handleBlur('email')} 
								name="email" 
								placeholder="Email" 
							/>
							<AuthInput
								value={password}
								touched={touched}
								error={validationError}
								handleChange={this.handleChange} 
								handleBlur={this.handleBlur('password')} 
								name="password"
								type="password" 
								placeholder="Password" 
							/>
							<AuthInput
								value={confirmPassword}
								touched={touched}
								error={validationError}
								handleChange={this.handleChange} 
								handleBlur={this.handleBlur('confirmPassword')} 
								name="confirmPassword" 
								type="password"
								placeholder="Confirm Password"
							/>
							<Button title="Sign Up" type="submit" />
						</form>
					</main>
				</div>
				<div className="authAlt">
					Already have an account?
					<Link to="/auth/signin">Sign In</Link>
				</div>
			</div>
		);
	}
};

export default connect(null, { signup })(SignupForm);