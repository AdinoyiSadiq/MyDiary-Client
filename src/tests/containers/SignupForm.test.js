import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import SignupForm from '../../containers/SignupForm';
import Root from '../../root';

let wrapped;

beforeEach(() => {
  wrapped = mount(
    <Root>
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <SignupForm />
      </MemoryRouter>
    </Root>,
  );
});

afterEach(() => wrapped.unmount());

describe('Signup UI', () => {
  describe('render features', () => {
    test('container should render as expected', () => {
      const tree = toJson(wrapped);
      expect(tree).toMatchSnapshot();
    });

    it('has five input fields and a button', () => {
      expect(wrapped.find('input').length).toEqual(5);
      expect(wrapped.find('button').length).toEqual(1);
    });
  });

  describe('when typing into fields', () => {
    beforeEach(() => {
      wrapped.find('input').first().simulate('change', { target: { id: 'firstname', value: 'john' } });
      wrapped.find('input').at(1).simulate('change', { target: { id: 'lastname', value: 'doe' } });
      wrapped.find('input').at(2).simulate('change', { target: { id: 'email', value: 'johndoe@email.com' } });
      wrapped.find('input').at(3).simulate('change', { target: { id: 'password', value: 'password' } });
      wrapped.find('input').at(4).simulate('change', { target: { id: 'confirmPassword', value: 'password' } });
      wrapped.update();
    });

    it('firstname field has been filled', () => {
      expect(wrapped.find('input').first().prop('value')).toEqual('john');
    });

    it('lastname field has been filled', () => {
      expect(wrapped.find('input').at(1).prop('value')).toEqual('doe');
    });

    it('email field has been filled', () => {
      expect(wrapped.find('input').at(2).prop('value')).toEqual('johndoe@email.com');
    });

    it('password field has been filled', () => {
      expect(wrapped.find('input').at(3).prop('value')).toEqual('password');
    });

    it('confirmPassword field has been filled', () => {
      expect(wrapped.find('input').at(4).prop('value')).toEqual('password');
    });

    it('empties fields when form is submitted', () => {
      wrapped.find('form').simulate('submit');
      wrapped.update();

      expect(wrapped.find('input').first().prop('value')).toEqual('');
      expect(wrapped.find('input').at(1).prop('value')).toEqual('');
      expect(wrapped.find('input').at(2).prop('value')).toEqual('');
      expect(wrapped.find('input').at(3).prop('value')).toEqual('');
      expect(wrapped.find('input').at(4).prop('value')).toEqual('');
    });
  });

  describe('return empty field when form is submitted', () => {
    it('empties fields when form is submitted', () => {
      wrapped.find('form').simulate('submit');
      wrapped.update();

      expect(wrapped.find('input').first().prop('value')).toEqual('');
      expect(wrapped.find('input').at(1).prop('value')).toEqual('');
      expect(wrapped.find('input').at(2).prop('value')).toEqual('');
      expect(wrapped.find('input').at(3).prop('value')).toEqual('');
      expect(wrapped.find('input').at(4).prop('value')).toEqual('');
    });
  });
});

describe('Signup client-side validation', () => {
  beforeEach(() => {
    wrapped.find('input').first().simulate('change', { target: { id: 'firstname', value: '' } });
    wrapped.find('input').at(1).simulate('change', { target: { id: 'lastname', value: '' } });
    wrapped.find('input').at(2).simulate('change', { target: { id: 'email', value: '' } });
    wrapped.find('input').at(3).simulate('change', { target: { id: 'password', value: '' } });
    wrapped.find('input').at(4).simulate('change', { target: { id: 'confirmPassword', value: '' } });
    wrapped.update();
  });

  it('shows error messages if both fields are empty', () => {
    wrapped.find('form').simulate('submit');
    wrapped.update();

    expect(wrapped.find('.authWrapper').first().find('.error').text()).toEqual('Please enter your first name');
    expect(wrapped.find('.authWrapper').at(1).find('.error').text()).toEqual('Please enter your last name');
    expect(wrapped.find('.authWrapper').at(2).find('.error').text()).toEqual('Please enter a valid email.');
    expect(wrapped.find('.authWrapper').at(3).find('.error').text()).toEqual('Please enter a valid password.');
    expect(wrapped.find('.authWrapper').at(4).find('.error').text()).toEqual('Please confirm your password.');
  });
});