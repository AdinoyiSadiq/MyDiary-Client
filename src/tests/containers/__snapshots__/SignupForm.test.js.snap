// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup UI render features container should render as expected 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(SignupForm)>
          <SignupForm
            signup={[Function]}
          >
            <div>
              <div
                className="authContent signup"
              >
                <main>
                  <div
                    className="authLogo"
                  >
                    MyDiary
                  </div>
                  <div
                    className="title"
                  >
                    Create your MyDiary account
                  </div>
                  <hr />
                  <form
                    className="signupForm"
                    onSubmit={[Function]}
                  >
                    <AuthInput
                      error={
                        Object {
                          "confirmPassword": "Please confirm your password.",
                          "email": "Please enter a valid email.",
                          "firstname": "Please enter your first name",
                          "lastname": "Please enter your last name",
                          "password": "Please enter a valid password.",
                          "status": true,
                        }
                      }
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      name="firstname"
                      placeholder="First name"
                      touched={
                        Object {
                          "confirmPassword": false,
                          "email": false,
                          "firstname": false,
                          "lastname": false,
                          "password": false,
                        }
                      }
                      value=""
                    >
                      <div
                        className="authWrapper"
                      >
                        <input
                          id="firstname"
                          onBlur={[Function]}
                          onChange={[Function]}
                          placeholder="First name"
                          type="text"
                          value=""
                        />
                        <div
                          className="error"
                        />
                      </div>
                    </AuthInput>
                    <AuthInput
                      error={
                        Object {
                          "confirmPassword": "Please confirm your password.",
                          "email": "Please enter a valid email.",
                          "firstname": "Please enter your first name",
                          "lastname": "Please enter your last name",
                          "password": "Please enter a valid password.",
                          "status": true,
                        }
                      }
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      name="lastname"
                      placeholder="Last name"
                      touched={
                        Object {
                          "confirmPassword": false,
                          "email": false,
                          "firstname": false,
                          "lastname": false,
                          "password": false,
                        }
                      }
                      value=""
                    >
                      <div
                        className="authWrapper"
                      >
                        <input
                          id="lastname"
                          onBlur={[Function]}
                          onChange={[Function]}
                          placeholder="Last name"
                          type="text"
                          value=""
                        />
                        <div
                          className="error"
                        />
                      </div>
                    </AuthInput>
                    <AuthInput
                      error={
                        Object {
                          "confirmPassword": "Please confirm your password.",
                          "email": "Please enter a valid email.",
                          "firstname": "Please enter your first name",
                          "lastname": "Please enter your last name",
                          "password": "Please enter a valid password.",
                          "status": true,
                        }
                      }
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      name="email"
                      placeholder="Email"
                      touched={
                        Object {
                          "confirmPassword": false,
                          "email": false,
                          "firstname": false,
                          "lastname": false,
                          "password": false,
                        }
                      }
                      value=""
                    >
                      <div
                        className="authWrapper"
                      >
                        <input
                          id="email"
                          onBlur={[Function]}
                          onChange={[Function]}
                          placeholder="Email"
                          type="text"
                          value=""
                        />
                        <div
                          className="error"
                        />
                      </div>
                    </AuthInput>
                    <AuthInput
                      error={
                        Object {
                          "confirmPassword": "Please confirm your password.",
                          "email": "Please enter a valid email.",
                          "firstname": "Please enter your first name",
                          "lastname": "Please enter your last name",
                          "password": "Please enter a valid password.",
                          "status": true,
                        }
                      }
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      name="password"
                      placeholder="Password"
                      touched={
                        Object {
                          "confirmPassword": false,
                          "email": false,
                          "firstname": false,
                          "lastname": false,
                          "password": false,
                        }
                      }
                      type="password"
                      value=""
                    >
                      <div
                        className="authWrapper"
                      >
                        <input
                          id="password"
                          onBlur={[Function]}
                          onChange={[Function]}
                          placeholder="Password"
                          type="password"
                          value=""
                        />
                        <div
                          className="error"
                        />
                      </div>
                    </AuthInput>
                    <AuthInput
                      error={
                        Object {
                          "confirmPassword": "Please confirm your password.",
                          "email": "Please enter a valid email.",
                          "firstname": "Please enter your first name",
                          "lastname": "Please enter your last name",
                          "password": "Please enter a valid password.",
                          "status": true,
                        }
                      }
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      name="confirmPassword"
                      placeholder="Confirm Password"
                      touched={
                        Object {
                          "confirmPassword": false,
                          "email": false,
                          "firstname": false,
                          "lastname": false,
                          "password": false,
                        }
                      }
                      type="password"
                      value=""
                    >
                      <div
                        className="authWrapper"
                      >
                        <input
                          id="confirmPassword"
                          onBlur={[Function]}
                          onChange={[Function]}
                          placeholder="Confirm Password"
                          type="password"
                          value=""
                        />
                        <div
                          className="error"
                        />
                      </div>
                    </AuthInput>
                    <Button
                      title="Sign Up"
                      type="submit"
                    >
                      <button
                        type="submit"
                      >
                        Sign Up
                      </button>
                    </Button>
                  </form>
                </main>
              </div>
              <div
                className="authAlt"
              >
                Already have an account?
                <Link
                  replace={false}
                  to="/auth/signin"
                >
                  <a
                    href="/auth/signin"
                    onClick={[Function]}
                  >
                    Sign In
                  </a>
                </Link>
              </div>
            </div>
          </SignupForm>
        </Connect(SignupForm)>
      </Router>
    </MemoryRouter>
  </Provider>
</_default>
`;
